engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java

```diff
@@ -539,6 +539,8 @@ private boolean isLightVisible(Vector3f positionViewSpace, LightComponent compon
 
     @Override
     public boolean isHeadUnderWater() {
+        // TODO: Making this as a subscribable value especially for node "ChunksRefractiveReflectiveNode",
+        // TODO: glDisable and glEnable state changes on that node will be dynamically added/removed based on this value.
```

Comment:
Sorry, I misread the comment. I now understand you were referring to adding/removing StateChange instances. We can try. Perhaps it won't be much of a performance hit. Certainly we should try to have no instantiation occurring during the switches. I.e. have everything ready at initialization and then only change the content of desiredStateChanges set. 

That been said changes in the set at this stage will always trigger a refresh of the whole intermediate and task lists. I'd be surprised if this doesn't have any impact. For this reason I suspect that whatever happens when the head is underwater might have to be handled with an if block within a Node.process() method, just to keep everything very responsive.